### PR TITLE
Hist benchmark fill

### PR DESCRIPTION
**Pull request by claireguyot**

### CODE REVIEW

File: root/hist/hist/THistFillBenchmarks.cxx [hunk 1]
```diff
@@ -16,69 +13,227 @@ static void FillArguments(benchmark::internal::Benchmark *b)
          b->Args({i, j, j / 2});
 }
 
+// Create array for x coordinates
+double * MakeXCoorVec(int nbOfDataPoints)
```
Comment (Axel-Naumann) [inline root/hist/hist/THistFillBenchmarks.cxx]: Can you return `vector<double>`?

File: root/hist/hist/THistFillBenchmarks.cxx [hunk 2]
```diff
@@ -16,69 +13,227 @@ static void FillArguments(benchmark::internal::Benchmark *b)
...
+   double * vec = new double[nbOfDataPoints];
+   for (int j = 0; j < nbOfDataPoints; ++j) {
+      double d = static_cast< double >(j);
+      vec[j] = d;
+   }
+   return vec;
+}
+
+// Arrays of coordinates and weights
+double * xCoords = MakeXCoorVec(3000000);
```
Comment (Axel-Naumann) [inline root/hist/hist/THistFillBenchmarks.cxx]: These should be `vector<double>`

File: root/hist/hist/THistFillBenchmarks.cxx [hunk 3]
```diff
@@ -16,69 +13,227 @@ static void FillArguments(benchmark::internal::Benchmark *b)
...
+double * yCoords = MakeYCoorVec(3000000);
+double * weights = MakeWeightVec(3000000);
+
+// Histograms with different bin configurations for Fill
+TH2D hist1{"hist1", "", 10, 0., 1., 10 / 2, 0., 10.};
```
Comment (Axel-Naumann) [inline root/hist/hist/THistFillBenchmarks.cxx]: Can we test overflow (or underflow) as well? E.g. by starting all x axes at 0.1?

File: root/hist/hist/THistFillBenchmarks.cxx [hunk 4]
```diff
@@ -16,69 +13,227 @@ static void FillArguments(benchmark::internal::Benchmark *b)
...
-      for (int i = 0; i < nbOfDataPoints; ++i) {
-         double d = static_cast< double >(i);
-         hist.Fill(d / nbOfDataPoints, d / (nbOfDataPoints / 10));
-      }
-
-      // Fill with weights
-      for (int i = 0; i < nbOfDataPoints; ++i) {
-         double d = static_cast< double >(i);
-         hist.Fill(d / nbOfDataPoints, d / (nbOfDataPoints / 10), d);
+      if (nbOfBinsForFirstAxis == 10) {
```
Comment (Axel-Naumann) [inline root/hist/hist/THistFillBenchmarks.cxx]: All of those are `TH2D`s, so we could have a `map<int, map<int, TH2D*>>` that allows you to do `hist[nbOfBinsForFirstAxis][nbOfBinsForSecondAxis]->Fill...`? And then move the two `for (...nbOfDataPoints...)` loops into a separate function, such that you just call `Fill(hist[nbOfBinsForFirstAxis][nbOfBinsForSecondAxis])`? That would save many many lines of repeated code.

File: root/hist/hist/THistFillBenchmarks.cxx [hunk 5]
```diff
@@ -16,69 +15,115 @@ static void FillArguments(benchmark::internal::Benchmark *b)
...
+TH2D hist6{"hist6", "", 100, 0.1, 1., 1000 / 2, 0., 10.};
+TH2D hist7{"hist7", "", 1000, 0.1, 1., 10 / 2, 0., 10.};
+TH2D hist8{"hist8", "", 1000, 0.1, 1., 100 / 2, 0., 10.};
+TH2D hist9{"hist9", "", 1000, 0.1, 1., 1000 / 2, 0., 10.};
+
+std::map<int, std::map<int, TH2D*>> histFill = { { 10, { { 5, &hist1 }, { 50, &hist2 }, { 500, &hist3} } },
+                                    { 100, { { 5, &hist4 }, { 50, &hist5 }, { 500, &hist6 } } },
+                                    { 1000, { { 5, &hist7 }, { 50, &hist8 }, { 500, &hist9 } } } };
+
+static void FillHist(int nbOfDataPoints, int nbOfBinsForFirstAxis, int nbOfBinsForSecondAxis)
```
Comment (Axel-Naumann) [inline root/hist/hist/THistFillBenchmarks.cxx]: I'd have passe the `TH2D` in, instead of the indexes into the map. But that's details.

Comment (claireguyot) [reply root/hist/hist/THistFillBenchmarks.cxx]: Should I change it?

**Comment by vgvassilev** [comment]
We should make the various `hist2` configurable instead of defining them as local variables. This way we can test the scalability easier with the already available benchmarking primitives.

**Comment by Axel-Naumann** [comment]
This is temporary until we validated that pure `Fill` has no constant overhead (would be surprising), at which point we'll simply pick *one* configuration.

**Comment by vgvassilev** [comment]
> This is temporary until we validated that pure `Fill` has no constant overhead (would be surprising), at which point we'll simply pick _one_ configuration.

I understand, my proposal would be even better for this use-case. Please make it configurable, it will save you code and after you validate, you will change just one token to select the configuration which runs nightly. Then, locally, if you want to test how it scales beyond eg 1000 you can just configure the stepping of the `Range` and play with it.

**Comment by Axel-Naumann** [comment]
What is "configurable"? Do you have an example?

**Comment by vgvassilev** [comment]
Configurable, means parametrized by in terms of google benchmark. Here is something similar: https://github.com/root-project/rootbench/blob/master/root/math/genvector/Cartesian3DBenchmarks.cxx#L93

**Comment by Axel-Naumann** [comment]
We had that previously (and we have it now, to some extent) - but we want to move the histogram construction out of the benchmark, so now the state / parameter selects which histogram we profile.

As I said: this is not needed, it's temporary, the next version will not need these parameters at all. We don't want them.

File: root/hist/hist/THistFillBenchmarks.cxx [hunk 6]
```diff
@@ -3,8 +3,7 @@
 
 #include "benchmark/benchmark.h"
 
-
-
+#include <map>
```
Comment (Axel-Naumann) [inline root/hist/hist/THistFillBenchmarks.cxx]: Can you `#include <vector>`?

File: root/hist/hist/THistFillBenchmarks.cxx [hunk 7]
```diff
@@ -16,69 +15,115 @@ static void FillArguments(benchmark::internal::Benchmark *b)
...
-   double * weights = new double[nbOfDataPoints];
-
-   for (int i = 0; i < nbOfDataPoints; ++i) {
-      double d = static_cast< double >(i);
-      xCoords[i] = d / nbOfDataPoints;
-      yCoords[i] = d / (nbOfDataPoints / 10);
-      weights[i] = d;
-   }
-
    for (auto _ : state) {
```
Comment (Axel-Naumann) [inline root/hist/hist/THistFillBenchmarks.cxx]: Can you move the map lookup outside the for loop:
```
auto hist = histFillN[nbOfBinsForFirstAxis][nbOfBinsForSecondAxis];
for ...
   hist->FillN
```
That way we don't benchmark the map lookup as much :-)

**Comment by vgvassilev** [comment]
Yet, that temporary code grows in the next PR. 

I am pretty sure you want to know how the construction of the different histograms with different parameter behaves. Going to ranges will allow you to do that and make this benchmark closer to the good practice recommended by google.

**Comment by Axel-Naumann** [comment]
> how the construction of the different histograms with different parameter behaves

Yes, but not here. This is about filling, and we're convinced that one config is sufficient.

**Comment by vgvassilev** [comment]
So you are proposing temporary code to see in grafana and then select one config?

**Comment by Axel-Naumann** [comment]
That's the plan, yes. I tried to get that across with my comment at https://github.com/root-project/rootbench/pull/109#issuecomment-594375244

> This is temporary until we validated that pure Fill has no constant overhead (would be surprising), at which point we'll simply pick one configuration.

**Comment by vgvassilev** [comment]
That's a strange workflow but there is nothing else we can provide on the infrastructure side except logging in on the physical machine and testing by hand (which is not what I suggest).

**Pull request merged**